### ISSUE TITLE
Attributes order is not preserved

### ISSUE THREAD

**Issue by fabiospampinato** [cure53/DOMPurify#586]
### Background & Context

I'm using DOMPurify to sanitize a big table people may be editing in real time, since the sanitization is fairly slow I want to cache sanitization for unchanged table rows, and I thought I'd remember the outerHTML string of table rows that are sanitized, and if the table row to sanitize has an outerHTML that matches that already then I basically skip sanitizing.

### Bug

The problem is that I can't use outerHTML as my caching key essentially, because DOMPurify returns me nodes with a different, _albeit 100% equivalent_, outerHTML string.

### Code

Executing the following:

```js
(() => {

  const parser = new DOMParser ();
  const parse = html => Array.from ( parser.parseFromString ( html, 'text/html' ).body.childNodes );
  
  const purifier = require ( 'dompurify' )( window );
  const purify = node => purifier.sanitize ( node, { ADD_ATTR: ['target'], IN_PLACE: true, RETURN_DOM: true } );

  const html_in1 = '<a href="#" target="_blank"></a>';
  const dom_in1 = parse ( html_in1 )[0];
  const dom_out1 = purify ( dom_in1 );
  const html_out1 = dom_out1.outerHTML;
  
  console.log(html_in1);
  console.log(html_out1);

  console.log('-------------');

  const html_in2 = '<a target="_blank" href="#"></a>';
  const dom_in2 = parse ( html_in2 )[0];
  const dom_out2 = purify ( dom_in2 );
  const html_out2 = dom_out2.outerHTML;
  
  console.log(html_in2);
  console.log(html_out2);

})()
```

Produces the following output:

```
<a href="#" target="_blank"></a>
<a target="_blank" href="#"></a>
-------------
<a target="_blank" href="#"></a>
<a href="#" target="_blank"></a>
```

Basically whatever order I use for my attributes DOMPurify flips it, it seems.

### Expected output

I'd expect the outerHTML value of a node that is actually entirely safe according to DOMPurify to be exactly the same before and after the sanitization, otherwise use cases like mine become unnecessarily complicated and slower to do.

I'd guess DOMPurify is possibly iterating over attributes from the end or something, it that's the case it'd be great if it instead iterated over them from the start so that the order is kept the same.

Also I don't understand why the NamedNodeMap object for attributes is being changed at all given that it's safe. That seems like a bug perhaps?

**Comment by fabiospampinato** [comment]
Apparently this issue got reported many times already.

IMO saying that different browsers provide different values for outerHTML misses the point a little, for use cases like mine it's important the the order stays consistent within an execution environment, it doesn't matter if in another browser the order is different, as long as it's consistent.

And obviously using DOMPurify causes the order to flip, since if I don't use it I always get the same  html string.

I don't really see why the attributes object must be mutated at all if no attribute actually needs changing at all, I think not mutating it would fix this issue for that common case at least, right? It may even be faster, which would be a nice bonus.

**Comment by cure53** [comment]
Wontfix. We're offering a sanitizer and not a markup manager.

**Comment by cure53** [comment]
Actually, it's not really only us doing much of the sorting, some browsers re-sort attributes themselves. Also, we learned very early on that ignoring that will cause bypasses - so sadly, I see no chance to change anything in this regard without risking XSS.

**Comment by fabiospampinato** [comment]
> Actually, it's not really only us doing much of the sorting, some browsers re-sort attributes themselves. 

Which browser returns different outerHTML strings for elements that are not mutated in any way? I'd like to test this out for myself.

> Also, we learned very early on that ignoring that will cause bypasses - so sadly, I see no chance to change anything in this regard without risking XSS.

I'm not sure what you are referring to, ignoring what? I'm not saying to ignore anything, I'm just saying if an attributes object is valid it shouldn't be mutated, are you saying that that can cause XSS? Because that seems impossible by definition to me.

**Comment by fabiospampinato** [comment]
> Check the test cases, you will see that browsers themselves change attributes a lot as they see it fit :)

Browsers change attributes at a whim without them being mutated in any way by the user? That seems unbelievable, I need to see this for myself. 

I can't spot something that I can execute from [here](https://github.com/cure53/DOMPurify/blob/main/test/fixtures/expect.js) to reproduce your claim, could you please point me at something more concrete? All those these cases I'd imagine go through DOMPurify, which I think is what's actually causing the issue here, so I wouldn't even know what to execute.

I can't reproduce the issue with something like the following under Chrome, nor Firefox, nor Safari:

```js
const target = document.body.outerHTML;
for ( let i = 0; i < 100; i++ ) {
 console.assert ( target === document.body.outerHTML )
}
```

**Comment by xmedeko** [comment]
@fabiospampinato  duplicate to #276 and #435 ?

**Comment by xmedeko** [comment]
Or workaround is to reverse the order in `afterSanitizeAttributes` hook:

```js
DOMPurify.addHook("afterSanitizeAttributes", (node) => {
  const { attributes } = node;
  if (!attributes)
    return;
  let l = attributes.length;
  while (l--) {
    const attr = attributes[l];
    const { name, namespaceURI, value } = attr;
    if (name === "is" && !namespaceURI) // non-removable "is" attribute
    	continue;
    if (namespaceURI) {
        node.removeAttributeNS(namespaceURI, name);
        node.setAttributeNS(namespaceURI, name, value);
    } else {
      node.removeAttribute(name);
      node.setAttribute(name, value);
    }
  }
});
```

**Comment by tv42** [comment]
Unfortunately the workaround posted breaks as soon as you use `forceKeepAttr`. That code path short circuits the remove-and-add-logic that otherwise reliably reversed the attribute order.

https://github.com/cure53/DOMPurify/blob/fcb9dbd9a935d91e1a087b5ee721da1c6b008790/src/purify.js#L1243-L1249

https://github.com/cure53/DOMPurify/blob/fcb9dbd9a935d91e1a087b5ee721da1c6b008790/src/purify.js#L1315-L1320
